public/app/plugins/panel/xychart/XYChartPanel2.tsx

```diff
@@ -0,0 +1,126 @@
+import React, { PureComponent } from 'react';
+import { PanelProps } from '@grafana/data';
+import { XYChartOptions } from './models.gen';
+import { ScatterHoverEvent, ScatterSeries } from './types';
+import {
+  LegendDisplayMode,
+  Portal,
+  UPlotChart,
+  UPlotConfigBuilder,
+  VizLayout,
+  VizLegend,
+  VizLegendItem,
+  VizTooltipContainer,
+} from '@grafana/ui';
+import { FacetedData } from '@grafana/ui/src/components/uPlot/types';
+import { prepData, prepScatter } from './scatter';
+import { config } from '@grafana/runtime';
+import { TooltipView } from './TooltipView';
+
+type Props = PanelProps<XYChartOptions>;
+type State = {
+  error?: string;
+  series: ScatterSeries[];
+  builder?: UPlotConfigBuilder;
+  facets?: FacetedData;
+  hover?: ScatterHoverEvent;
+};
+
+export class XYChartPanel2 extends PureComponent<Props, State> {
+  state: State = {
+    series: [],
+  };
+
+  componentDidMount() {
+    this.initSeries(); // also data
+  }
+
+  componentDidUpdate(oldProps: Props) {
+    const { options, data } = this.props;
+    const configsChanged = options !== oldProps.options || data.structureRev !== oldProps.data.structureRev;
```

Comment:
`options !== oldProps.options` - I'm not 100% sure about this one - ain't these two going to be the same in situation when deep options change, i.e. thresholds or value mappings config? My gut feeling is it's gonna be indicated by the structureRev change, but still quite complex. Think this wasn't that visible with the hooks invalidation, which makes me like class component approach better in this case.